### PR TITLE
[FIXED JENKINS-26452] Prevent NPE if no publisher in conditional step.

### PR DESCRIPTION
**Pull request by christ66**
[JENKINS-26452](https://issues.jenkins-ci.org/browse/JENKINS-26452)

If there are no publishers added to the flexible publish plugin, then the plugin can throw a NPE when saving, executing a build, and restarting Jenkins will fail to load the job.

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkins_ci/plugins/flexible_publish/ConditionalPublisher.java [hunk 1]
```diff
@@ -137,9 +141,19 @@ public BuildStep getPublisher() {
...
 
     public List<BuildStep> getPublisherList() {
-        return publisherList != null?publisherList:Collections.<BuildStep>emptyList();
-    }
+        if (publisherList != null) {
+            // We also need to guard against any null values in the publisherList
+            // This can happen if the projects publisher list was removed, returning
+            // a list with one null element.
+            if (publisherList.contains(null)) {
+                publisherList.remove(null);
```
Comment (jglick) [inline src/main/java/org/jenkins_ci/plugins/flexible_publish/ConditionalPublisher.java]: Better done in `readResolve`.

**Comment by jglick** [comment]
While the changes here seem fine as a robustness improvement, I feel like something is missing. How could the UI form be submitting a list with null elements to begin with? The reproduction step

> 3. Save the project _without_ any conditional publishers.

should just result in an empty list if standard form binding is being used, and that should be harmless.

**Comment by christ66** [comment]
The problem was the `ConditionalPublisher.newInstance` method could have added a null value to the list. I also made modifications to not throw any exception, and just have an empty list instead.

**Comment by ikedam** [comment]
Looks almost perfect to me. Thanks.
* Please revert unrelated whitespace changes. You can check them in the "Files changed" tab.
* I created JENKINS-26452 as JENKINS-19985 is for another issue. Would you work on that?
    https://issues.jenkins-ci.org/browse/JENKINS-26452

**Comment by jglick** [comment]
In the long term it would be better to get rid of the `newInstance` override entirely, but in the meantime :+1: other than the gratuitous whitespace changes.

**Comment by christ66** [comment]
Sorry about the whitespace changes. I thought I had fixed my IDE to not make these changes.

File: src/test/java/org/jenkins_ci/plugins/flexible_publish/ConfigurationTest.java [hunk 2]
```diff
@@ -87,6 +90,29 @@ public void testWithoutPublishers() throws Exception {
...
+      ConditionalPublisher conditionalPublisher = new ConditionalPublisher(new AlwaysRun(),
+              Collections.<BuildStep>singletonList(null),
+              null,
+              false,
+              null,
+              null);
+      FlexiblePublisher fp = new FlexiblePublisher(Arrays.asList(conditionalPublisher));
+      p.getPublishersList().add(fp);
+      p.save();
+      jenkins.reload();
```
Comment (jglick) [inline src/test/java/org/jenkins_ci/plugins/flexible_publish/ConfigurationTest.java]: BTW another way of doing this is to use `@LocalData` with a prepared example of historically broken configuration.

**Comment by jglick** [comment]
:+1:

**Comment by ikedam** [comment]
Thanks for the pull request!
I released 0.14.1, which will be available in a day.

**Comment by ikedam** [comment]
Oh... I had to release as 0.13.1 (that means fix for 0.13)...
I mistook!

**Pull request merged**